### PR TITLE
Plugin framework: add ability to install other plugins to the plugin API

### PR DESCRIPTION
**Pull request by niklabh**
#### Summary

Plugin framework: add ability to upload other plugins to the plugin API
- added `InstallPlugin(file io.Reader, replace bool) (*model.Manifest, *model.AppError)` method to plugin API
#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/10289

### CODE REVIEW

File: app/plugin_api.go [hunk 1]
```diff
@@ -655,6 +657,20 @@ func (api *PluginAPI) GetPluginStatus(id string) (*model.PluginStatus, *model.Ap
 	return api.app.GetPluginStatus(id)
 }
 
+func (api *PluginAPI) UploadPlugin(file io.Reader, replace bool) (*model.Manifest, *model.AppError) {
+	if !*api.app.Config().PluginSettings.Enable || !*api.app.Config().PluginSettings.EnableUploads {
+		return nil, model.NewAppError("uploadPlugin", "app.plugin.upload_disabled.app_error", nil, "", http.StatusNotImplemented)
+	}
+
+	fileBuffer, err := ioutil.ReadAll(file)
```
Comment (crspeller) [inline app/plugin_api.go]: What is the purpose of reading the file into the buffer here?

Comment (niklabh) [reply app/plugin_api.go]: As per ticket specifications UploadPlugin method should take an io.Reader. InstallPluin api require a io.ReadSeeker. io.Reader does not implement seek(). To make this happen file needs to be read completely. So reading file in memory and creating a ReadSeeker using bytes.NewReader

Comment (crspeller) [reply app/plugin_api.go]: Ok That make sense.

**Comment by crspeller** [comment]
@lieut-data @hanzei Wondering about the naming here. Why is an API called `UploadPlugin` calling `InstallPlugin`.

@niklabh No fault of you. Implemented to the ticket. Thanks!

**Comment by hanzei** [comment]
My biggest reason was consistency with the REST API at the time writing this ticket (https://api.mattermost.com/#tag/plugins%2Fpaths%2F~1plugins%2Fpost). But have broken consistency already by adding `Install plugin from URL`(https://api.mattermost.com/#tag/plugins%2Fpaths%2F~1plugins~1install_from_url%2Fpost)

The other point I would make for `Upload` is, that `Install` indicates that the plugin in running after the method call, what is not the case. EnablePlugin() needs to be called after this method to enable it. See https://github.com/mattermost/mattermost-server/pull/10697

I'm 0/5 what the should use. But if we go with `Install`, we should also use this term in the product.

**Comment by niklabh** [comment]
Any changes required here?

**Comment by hanzei** [comment]
@niklabh @crspeller was out of office last week and is this week. Hence it will take some time for the review.

In the mean time it would be great if you would add a unit test for this.

**Comment by niklabh** [comment]
test setup is failing. Can anyone help here.

**Comment by lieut-data** [comment]
Not sure why the build failed -- I've tried kicking it again.

**Comment by lieut-data** [comment]
Asking @ali-farooq0 and @iomodo to review given our pending work on plugin signing and to ensure Toolkit alignment.

**Comment by niklabh** [comment]
> Not sure why the build failed -- I've tried kicking it again.

Build is passing now

File: plugin/api.go [hunk 2]
```diff
@@ -557,6 +559,12 @@ type API interface {
 	// Minimum server version: 5.6
 	GetPluginStatus(id string) (*model.PluginStatus, *model.AppError)
 
+	// UploadPlugin will upload another plugin with tar.gz file.
+	// Previous version will be replaced on replace true.
+	//
+	// Minimum server version: 5.6
```
Comment (ali-farooq0) [inline plugin/api.go]: i believe this should be `5.18`

File: app/plugin_api.go [hunk 3]
```diff
@@ -655,6 +657,20 @@ func (api *PluginAPI) GetPluginStatus(id string) (*model.PluginStatus, *model.Ap
...
+	if err != nil {
+		return nil, model.NewAppError("uploadPlugin", "api.plugin.upload.file.app_error", nil, "", http.StatusBadRequest)
+	}
+	fileReader := bytes.NewReader(fileBuffer)
+
+	return api.app.InstallPlugin(fileReader, replace)
```
Comment (ali-farooq0) [inline app/plugin_api.go]: could just single line it `return api.app.InstallPlugin(bytes.NewReader(fileBuffer), replace)`

File: app/plugin_api_test.go [hunk 4]
```diff
@@ -684,6 +684,30 @@ func TestPluginAPIGetPlugins(t *testing.T) {
 	assert.Equal(t, pluginManifests, plugins)
 }
 
+func TestPluginAPIUploadPlugin(t *testing.T) {
```
Comment (ali-farooq0) [inline app/plugin_api_test.go]: can we expand on this to include a test that successfully uploads a plugin?

Comment (niklabh) [reply app/plugin_api_test.go]: done

File: plugin/api.go (hunk 2 above)
Comment (niklabh) [reply plugin/api.go]: done

**Comment by ali-farooq0** [comment]
@niklabh just a reminder that always try and avoid force pushing if a PR is currently being reviewed.
See https://developers.mattermost.com/blog/submitting-great-prs/#4-avoid-force-pushing

**Comment by niklabh** [comment]
noted

File: app/plugin_api_test.go (hunk 4 above)
Comment (ali-farooq0) [reply app/plugin_api_test.go]: doesn't look like this was updated?

**Comment by niklabh** [comment]
Added successful plugin test

**Comment by niklabh** [comment]
Anything else needed to be done here

**Comment by ali-farooq0** [comment]
Any outstanding concerns for this PR @crspeller?

**Comment by crspeller** [comment]
@ali-farooq0 @hanzei Just wondering about the name here. Someone should figure out which one we want and make it consistant.

**Comment by hanzei** [comment]
@ali-farooq0 What do you think about talking about this in a meeting?

**Comment by ali-farooq0** [comment]
> @ali-farooq0 @hanzei Just wondering about the name here. Someone should figure out which one we want and make it consistant.
> 
> @ali-farooq0 Can you handle that?

@crspeller Will do.



> @ali-farooq0 What do you think about talking about this in a meeting?

Sounds good @hanzei

**Comment by ali-farooq0** [comment]
@niklabh Discussed it with the team and consensus was to use `InstallPlugin` as opposed to `UploadPlugin`. Do you mind making the change so we can get this merged in?

cc: @hanzei, @crspeller

**Comment by niklabh** [comment]
on it

**Comment by niklabh** [comment]
done

**Pull request merged**